sched/Kconfig

```diff
@@ -975,7 +975,7 @@ config SCHED_INSTRUMENTATION_SPINLOCKS
 config SCHED_INSTRUMENTATION_SYSCALL
 	bool "System call monitor hooks"
 	default n
-	depends on LIB_SYSCALL && ARCH_HAVE_SYSCALL_HOOKS
+	depends on ARCH_HAVE_SYSCALL_HOOKS
```

Comment:
This seems incorrect.  It should not be possible to enable syscall instrumentation hooks if system calls are not enabled.  The option should never be visible to the user if it does not function.

Please revert to the original version or provide explanation.  If CONFIG_LIB_SYSCALL is not selected, then CONFIG_INSTRUMENTATION_SYSCALL must not be selectable or visiable to the user.

Reply:
In my understanding, CONFIG_LIB_SYSCALL is defined in PROTECTED and KERNEL build and not in FLAT build.
I have made CONFIG_INSTRUMENTATION_SYSCALL feature available in all three build types and it is the reason I removed LIB_SYSCALL.

Reply:
I think the problem will resolve automatically If we decouple hook and syscall.

Reply:
@patacongo since the syscall isn't enabled in the many projects, but it's very useful to hook 'syscall' in all situation, @YuuichiNakamura utilize the option '--wrap' provided by ld to hook in this case. This link has more information:
https://github.com/YuuichiNakamura/nuttx-task-tracer-doc/blob/master/NuttXTaskTracerInternals.md
So you can think that hook can decouple from syscall totally with --wrap.

Reply:
> In my understanding, CONFIG_LIB_SYSCALL is defined in PROTECTED and KERNEL build and not in FLAT build.
> I have made CONFIG_INSTRUMENTATION_SYSCALL feature available in all three build types and it is the reason I removed LIB_SYSCALL.

Incorrect!  Your understanding is wrong.  System calls are also used in the FLAT build to support external modules.  The configuration change you made must be reverted.  Please restore the dependency on LIB_SYSCALL.  That is required and will make the option operate correctly in all three build modes.

This change cannot be merged unless the dependency is restored.

Reply:
> @patacongo since the syscall isn't enabled in the many projects, but it's very useful to hook 'syscall' in all situation, @YuuichiNakamura utilize the option '--wrap' provided by ld to hook in this case. This link has more information:
> https://github.com/YuuichiNakamura/nuttx-task-tracer-doc/blob/master/NuttXTaskTracerInternals.md
> So you can think that hook can decouple from syscall totally with --wrap.

I don't fully agree.  The menu item must not be visible if CONFIG_LIB_SYCALL is not enabled.  That is the requirement.  The existing dependency is the simplest and surest way to assure that requirement.

But if the menu selection is never presented if CONFIG_LIB_SYSCALL is not selected through some other mechanism, then we can consider that.  The objective is to not burden the end user with many configuration options that do not do anything.  The selection must not be present or visible in any way if CONFIG_LIB_SYSCALL is not selected.

Reply:
This absolutely must be reverted!!!!!!!!!!!!!!!!!!!!!!

In the FLAT build, CONFIG_LIB_SYSCALL may or may not be defined.  If this CONFIG_LIB_SYSCALL is NOT defined, then this selection will be present in the menus.  THAT IS ABSOLUTELY WRONG.  There should be no invalid menu selection visible.  This is necessary to keep the complexity of the configuration menus down.

And the logic is wrong AND IS A BUG.  If some does select CONFIG_SCHED_INSTRUMENTATION_SYSCALL when CONFIG_LIB_SYSCALL is *not* defined then the build will fail.  THAT CANNOT BE PERMITTED.  We must no permit invalid configuration options that can cause build errors from being selected.

THIS MUST BE REVERTED!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Reply:
Yes, in the FLAT build, CONFIG_LIB_SYSCALL may or may not be defined, but  SCHED_INSTRUMENTATION_SYSCALL CAN STILL work as expect even without CONFIG_LIB_SYSCALL. That is why @YuuichiNakamura remove the dependence.

Reply:
I think that there there still must be an option to enable the instrumentation in the syscall stubs.  We have to be able to generate stubs without the instrumentation.  But yes, if it is inside of the syscall stubs, then that part of the logic does not need to be conditioned on CONFIG_LIB_SYSCALL.

But how about the generation of the event handling logic itself?

Some of this comes back to enabling static vs. runtime instrumentation.  The static controls are essential for resource limited and for systems with extremely high data rate events.

Better terminology... we need "frontend" and possibly "backend" controls:  There are basically two steps:

frontend:  instrumentation calls are received and data is buffered in memory.
backend:  The application reads the data from memory.

We need frontend controls to manage how much data is buffered and the rate of buffering.  That was accomplished in the past through static configuration settings.  If you want to avoid static settings, then you will need other controls to manage the rate that data is buffered.

Reply:
You still need the CONFIG_LIB_SYSCALL dependency to hide the menu option.  It is not needed for the instrumentation, but it is needed to keep garbage selections  out of the menus.  Otherwise the menu selection will always be available even  when it is not appropriate.

Reply:
> I think that there there still must be an option to enable the instrumentation in the syscall stubs. We have to be able to generate stubs without the instrumentation. But yes, if it is inside of the syscall stubs, then that part of the logic does not need to be conditioned on CONFIG_LIB_SYSCALL.
> 

Yes, it's controlled by SCHED_INSTRUMENTATION_SYSCALL. CONFIG_LIB_SYSCALL enable syscall, SCHED_INSTRUMENTATION_SYSCALL enable instrumentation. But since the instrumentation can work without syscall, the dependence has to be removed.

> But how about the generation of the event handling logic itself?
> 
> Some of this comes back to enabling static vs. runtime instrumentation. The static controls are essential for resource limited and for systems with extremely high data rate events.
> 
> Better terminology... we need "frontend" and possibly "backend" controls: There are basically two steps:
> 
> frontend: instrumentation calls are received and data is buffered in memory.
> backend: The application reads the data from memory.
> 
> We need frontend controls to manage how much data is buffered and the rate of buffering. That was accomplished in the past through static configuration settings. If you want to avoid static settings, then you will need other controls to manage the rate that data is buffered.

@YuuichiNakamura already support the filter and document here:
https://github.com/YuuichiNakamura/nuttx-task-tracer-doc/blob/master/NuttXTaskTracer.md
This filter is done at the runtime(I guess inside the sched_trace.c/sched_note.c driver). So we can:
1.Turn on/off the whole syscall instrumentation by SCHED_INSTRUMENTATION_SYSCALL
2.Turn on/off the individual syscall from command line at runtime
The 3rd approach(static filter) only instrument a specific set of syscall(e.g. provide the cut down syscall.csv). But I wonder whether the static filter is required if we already have more powerful runtime filter.

Reply:
> @YuuichiNakamura already support the filter and document here:
> https://github.com/YuuichiNakamura/nuttx-task-tracer-doc/blob/master/NuttXTaskTracer.md
> This filter is done at the runtime(I guess inside the sched_trace.c/sched_note.c driver). So we can:
> 1.Turn on/off the whole syscall instrumentation by SCHED_INSTRUMENTATION_SYSCALL
> 2.Turn on/off the individual syscall from command line at runtime
> The 3rd approach(static filter) only instrument a specific set of syscall(e.g. provide the cut down syscall.csv). But I wonder whether the static filter is required if we already have more powerful runtime filter.

Of course, that #2) CANNOT work for the hardware based solution since there is no driver in the hardware solution.